desktop/notification/gtk.go

```diff
@@ -0,0 +1,100 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package notification
+
+import (
+	"time"
+
+	"github.com/godbus/dbus"
+)
+
+const (
+	gtkBusName    = "org.gtk.Notifications"
+	gtkObjectPath = "/org/gtk/Notifications"
+	gtkInterface  = "org.gtk.Notifications"
+)
+
+type gtkBackend struct {
+	conn      *dbus.Conn
+	manager   dbus.BusObject
+	desktopID string
+	firstUse  time.Time
+}
+
+// TODO: support actions via session agent.
+var newGtkBackend = func(conn *dbus.Conn, desktopID string) (NotificationManager, error) {
+	// If the D-Bus service is not already running, assume it is
+	// not available.
+	var owner string
+	if err := conn.BusObject().Call("org.freedesktop.DBus.GetNameOwner", 0, gtkBusName).Store(&owner); err != nil {
+		return nil, err
+	}
+
+	b := &gtkBackend{
+		conn:      conn,
+		manager:   conn.Object(gtkBusName, gtkObjectPath),
+		desktopID: desktopID,
+	}
+	return b, nil
+}
+
+func gtkPriority(priority Priority) string {
+	switch priority {
+	case PriorityLow:
+		return "low"
+	case PriorityNormal:
+		return "normal"
+	case PriorityHigh:
+		return "high"
+	case PriorityUrgent:
+		return "urgent"
+	default:
+		return "normal" // default to normal
+	}
+}
+
+func (srv *gtkBackend) IdleDuration() time.Duration {
+	return timeNow().Sub(srv.firstUse)
```

Comment:
This seems to use a very different logic from the fdo backend: there, it looks like we are returning the time elapsed since the *last* interaction, here the time elapsed since the *first* one.

Reply:
That is correct. The intent of the IdleDuration() method is to help determine when it is safe for the snap session agent to exit.

When using FDO notifications, we want to stay active as long as any of our posted notifications are still visible (so it is possible to withdraw or update them, or respond to users clicking on them).

For the GTK API, notification IDs are assigned by the client.  So it is possible to withdraw/update notifications posted by a previous instance of the session agent. And activating a notification is handled via a D-Bus method call to the application that posted it, which may involve D-Bus activation if the session agent isn't currently running.

So there isn't really any reason for the GTK backend to keep the session agent alive.

Reply:
thanks for that explanation, makes a lot of sense